src/object.rs

```diff
@@ -751,24 +751,25 @@ pub trait SecretKeyFactory {
         mut data: Vec<u8>,
         template: &[CK_ATTRIBUTE],
     ) -> KResult<Object> {
-        let mut attrs = template.to_vec();
-        match attrs.iter().position(|x| x.type_ == CKA_VALUE_LEN) {
-            Some(idx) => {
-                let len = attrs[idx].to_ulong()?;
-                if (len as usize) < data.len() {
-                    unsafe { data.set_len(len as usize) };
-                }
-                if (len as usize) > data.len() {
-                    data.zeroize();
-                    return err_rv!(CKR_KEY_SIZE_RANGE);
+        let mut obj =
+            match template.iter().position(|x| x.type_ == CKA_VALUE_LEN) {
+                Some(idx) => {
+                    let len = template[idx].to_ulong()?;
+                    if (len as usize) < data.len() {
+                        unsafe { data.set_len(len as usize) };
+                    }
+                    if (len as usize) > data.len() {
+                        data.zeroize();
+                        return err_rv!(CKR_KEY_SIZE_RANGE);
+                    }
+                    let mut tmpl = template.to_vec();
+                    let _ = tmpl.swap_remove(idx);
```

Comment:
This removes the `CKA_VALUE_LEN` from the resulting object. But I would argue that it could be useful at times. It duplicates the information in the `CKA_VALUE`, if the value is readable, but otherwise there is no way for the caller to know if it is aes128, aes256 or key of different size.
So I think it should be kept in the object.

Reply:
Uhmmm I had the impression that you could not keep it, but it turns out you only can't specify it on object creation (key import), nothing is said about the other cases.
However I think we remove it also in case of key generation.
The idea was to have a consistent representation of all keys with CKA_VALUE_LEN always omitted.
But you make a good argument about knowing the key length.

I will create a separate PR where I make sure it is always added to the key objects, together with some other stuff I found.

Reply:
Followup bug in #32